enhancements/sig-architecture/224-event-based-manifestwork/README.md

```diff
@@ -0,0 +1,1042 @@
+
+## Motivation
+
+Manifestworks are used to deliver manifests from a hub cluster to managed clusters. As manifestwork adoption and managed cluster numbers increase, we encounter scaling issues due to excessive manifestwork creation on the hub. This adversely affects the performance of the hub apiserver. This proposal is to explore treating manifestworks as event messages instead of storing them in ETCD.
+
+By considering each cluster namespace as a separate topic or channel, managed cluster agents can subscribe to relevant manifestworks. This approach eliminates the need to store manifestworks on the hub cluster, as they are primarily consumed by controllers like the addon manager or manifestworkreplicaset controller. Additionally, this change allows other architectures, such as non-Kubernetes restful services with databases, to consume ManifestWorks.
+
+Moreover, with the adoption of the event-based mechanism, ManifestWorks become highly versatile and applicable in IoT or edge computing scenarios, in which resource constraints and intermittent connectivity are common, the event-driven approach allows ManifestWorks to be efficiently distributed and applied to edge clusters. This enhances the overall flexibility and adaptability of ManifestWorks, extending their utility beyond traditional Kubernetes environments.
+
+To facilitate easy adoption of the event-based mechanism, a client interface should be provided, enabling existing work agents or consumers to transition without extensive code changes.
+
+### Terminology
+
+- Source: The API consumer or component that utilizes this specification for ManifestWorks. It can be a controller on the hub cluster or a RESTful service handling resource requests. For RESTful service source, it often has a database to store the requested resources. Source should have a unique identifier to ensure its uniqueness and consistency, for example, a hub controller can generate a source ID by hashing the hub cluster URL and appending the controller name. Similarly, a RESTful service can select a unique name or generate a unique ID in the associated database for its source identification.
+- Work Agent: The controller that handles the deployment of requested resources on the managed cluster and status report to the source. Work agent should have a unique identifier to ensure its uniqueness and consistency. Usually, work agent is uniquely identified by a string that combines managed cluster name and agent name for consistency.
+- Broker: The message broker that handles the event transfer between the source and the work agent. It is represented by a message broker, such as MQTT or Kafka.
+
+### Goals
+
+- Establish a standardized event protocol for delivering manifestworks and returning status information.
+- Define a resync mechanism to handle failures and ensure synchronization.
```

Comment:
In the case where the agent is disconnected from the broker and the agent restarts, what would be the expected behavior of the agent?

Reply:
the sources/agents need to send the resync request to sync the resource status/spec

Reply:
If the broker is still disconnected, what would the behavior be?

Reply:
the sources/agents may need to retry connecting the broker periodically and we may also need a retry times for source/agents, the sources/agents should fail when the number of retries is reached